Duplicati/Library/Snapshots/NoSnapshotWindows.cs

```diff
@@ -93,12 +93,7 @@ public override System.IO.Stream OpenRead (string localPath)
         /// <param name='localFolderPath'>The folder to examinate</param>
         protected override string[] ListFiles (string localFolderPath)
         {
-            string[] tmp = SystemIO.IO_WIN.GetFiles(localFolderPath);
-            string[] res = new string[tmp.Length];
-            for(int i = 0; i < tmp.Length; i++)
-                res[i] = SystemIOWindows.StripUNCPrefix(tmp[i]);

```

Comment:
This seems to have implications beyond the fix. Is this required for the the folder listing functionality to work?
The methods are used throughout the code, especially in the backup code. I would think this fix would start recording files with UNC paths instead of their normalized form?

Reply:
Regarding this question:
> Is this required for the the folder listing functionality to work?

The change you highlight is intended to make the backup itself consistent.  Without this change, if you manually enter, for example, "`\\?\C:\Temp\`" in the source data, the resulting Duplicati backup data will contain an entry for the path "`\\?\C:\Temp\`", but the entries for subdirectories and files will not include the "`\\?\`" prefix; e.g., you might see "`C:\Temp\file.txt`".  The inconsistency in the backup then leaks through to the restore GUI that will show entries for both "`\\?\C:\Temp\`" and "`C:\Temp\`" in the file picker.  With this change, if someone manually enters "`\\?\C:\Temp\`", all the files and subdirectories will also be recorded with the prefix.  And the restore GUI then displays directory "`\\?\C:\Temp\`" and the appropriate files and subdirectories beneath it.

Regarding this question:
> The methods are used throughout the code, especially in the backup code. I would think this fix would start recording files with UNC paths instead of their normalized form?

Pull request #4256 changed `SystemIO.IO_WIN.GetFiles()` to essentially honor the form of the `path` argument in the results it returns.  If the `path` passed into `SystemIO.IO_WIN.GetFiles()` is itself _not_ prefixed with "`\\?\`", the code behaves like it did before #4256 and it strips the prefix from the results.  On the other hand, if `path` is already prefixed, `SystemIO.IO_WIN.GetFiles()` passes it straight to `Directory.GetFiles()` and the results of `SystemIO.IO_WIN.GetFiles()` will likewise be prefixed.

The code you highlight here changes `NoSnapshotWindows.ListFiles()` to _also_ honor the form of its path argument, `localFolderPath`.  If `localFolderPath` is not prefixed, the results will also not be prefixed.

For places in Duplicati that end up calling `NoSnapshotWindows.ListFiles()` with a non-prefixed path, Duplicati would continue to behave like it did before #4256 and before this pull request.  On the other hand, if there are places that call `NoSnapshotWindows.ListFiles()` with prefixed paths but were expecting non-prefixed results, that would be a problem.  And the same goes for the changes in #4256 and `SystemIO.IO_WIN.GetFiles()`.

I've done backups with non-prefixed paths with these changes; e.g., using the source data picker, and the entries stored in the backup are not prefixed, just like before these changes.  Also, the unit tests pass, though I don't know if the tests would be sensitive to this.

I'll take a look at the call tree for these functions and see how paths are obtained and how they are used to see if I can identify areas that might pose a problem.  If you have any specific things you can think of for me to test, I can do that, too.

Reply:
Just to follow up, I did look at the call tree for these functions and ran some basic experiments.  As far as I can tell, `\\?\` prefixed paths will only be recorded if the user provides `\\?\` prefixed paths, either through manually typing in paths this way via the GUI or by providing them on the command line.  That was the intent of these changes.  I have added another comment on this pull request for an alternative that would change Duplicati to _never_ store `\\?\` prefixed paths, if that would be preferred.